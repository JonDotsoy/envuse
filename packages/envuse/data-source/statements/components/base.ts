import { BufferCursor } from "../lib/buffer-cursor";
import { UnexpectedTokenError } from "../tdo/unexpected-token-error";
import { EventEmitter } from "events";
import util, { inspect, InspectOptions } from "util";
import { TypeNamesList } from "../tdo/type-names-list";

export type BaseType = {
  $type: TypeNamesList;
};

export type BaseExportType = {
  pos: number;
  end: number;
  children?: BaseExportType[];
};

export type BaseExportTypeJSON<T = TypeNamesList> = BaseExportType & {
  $type: T;
};

const elmPrefix = (elm: Base) => {
  let countPrefix = 0;
  let currentElement = elm;

  while (currentElement.parent) {
    countPrefix += 1;
    currentElement = currentElement.parent;
  }

  if (elm.parent) {
    return `${`  `.repeat(countPrefix)}`;
  }

  return "";
};

export const printBodyElement = (element: Base) => {
  const str = JSON.stringify(
    element.body.slice(element.pos, element.end).toString()
  );
  if (element.elementList.length > 1) {
    return str.length > 10 ? `${str.substring(0, 10)}...` : str;
  }
  return str;
};

export const printElement = (element: Base) =>
  `${element.toObjectName()} (${element.pos}, ${
    element.end
  }): ${printBodyElement(element)}`;

type BodyAssigned<T> = T extends { propsMutable: infer R }
  ? R extends keyof T
    ? Partial<Pick<T, R>>
    : {}
  : {};

type BaseEvents = {
  create_element: (element: Base) => void;
};

type ArgsType<T> = T extends (...args: infer R) => void ? R : [];

export type RejectOptions = {
  message?: string;
  position?: number;
};

export abstract class Base {
  abstract $type: TypeNamesList;

  #events = new EventEmitter();

  #parent: Base | null = null;

  end: number = this.pos;
  children: Base[] = [];
  oChildren: Base[] = [];
  #elementList: Base[] = [this];

  raw = Buffer.from([]);
  _raw: string = "";
  #filename: string | null;
  #bufferCursor: BufferCursor;
  #body: Buffer;

  constructor(
    filename: string | null,
    body: Buffer,
    public pos: number,
    bufferCursor = new BufferCursor(body)
  ) {
    this.#filename = filename;
    this.#bufferCursor = bufferCursor;
    this.#body = body;
  }

  get parent() {
    return this.#parent;
  }

  toObjectName() {
    return this.constructor.name;
  }

  toString() {
    return printElement(this);
  }

  [util.inspect.custom](depth: number, inspectOptions: InspectOptions) {
    let out = `${printElement(this)}`;

    const oChildren = this.oChildren.filter((c) => {
      if (inspectOptions.showHidden === true) return true;

      // Hidden elements
      const hiddenElements: TypeNamesList[] = ["Space", "SpaceNewLine"];

      return !hiddenElements.includes(c.$type);
    });

    if (depth <= 0) {
      return `${out}${oChildren.length ? `\n  ...` : ``}`;
    }

    // console.log(inspectOptions)
    if (oChildren.length) {
      for (const child of oChildren) {
        out += `\n`;
        out += inspect(child, { ...inspectOptions, depth: depth - 1 })
          .replace(/(^)/, "  ")
          .replace(/\n/g, "\n  ");
      }
      // out += `\n${this.oChildren.map(el => el[util.inspect.custom](depth, inspectOptions)).join("\n")}`;
    }
    return out;
  }

  on<T extends keyof BaseEvents>(event: T, listener: BaseEvents[T]) {
    this.#events.on(event, listener);
    return () => {
      this.removeListener(event, listener);
    };
  }

  removeListener<T extends keyof BaseEvents>(
    event: T,
    listener: BaseEvents[T]
  ) {
    this.#events.removeListener(event, listener);
  }

  private emit<T extends keyof BaseEvents>(
    event: T,
    ...args: ArgsType<BaseEvents[T]>
  ) {
    this.#events.emit(event, ...args);
  }

  get filename() {
    return this.#filename;
  }
  get bufferCursor() {
    return this.#bufferCursor;
  }
  get body() {
    return this.#body;
  }
  get elementList() {
    return this.#elementList;
  }

  private load() {
    const unsubscribeForward = this.bufferCursor.on("forward", () => {
      this.end = this.bufferCursor.position;
    });
    const unsubscribeCreateElement = this.on("create_element", (element) => {
      this.#elementList.push(element);
    });
    this.prepare(this.bufferCursor);
    unsubscribeForward();
    unsubscribeCreateElement();
    return this;
  }

  createElement<T extends Base>(
    Comp: {
      new (
        filename: string | null,
        body: Buffer,
        pos: number,
        bufferCursor?: BufferCursor
      ): T;
    },
    assign?: BodyAssigned<T>
  ) {
    const comp = new Comp(
      this.filename,
      this.body,
      this.bufferCursor.position,
      this.bufferCursor
    );

    this.emit("create_element", comp);
    Base.createElement(comp, assign);

    comp.#parent = this;

    const [, ...elements] = comp.elementList;

    this.elementList.push(...elements);
    this.oChildren.push(comp);

    return comp;
  }

  static createElement<T extends Base>(comp: T, assign?: BodyAssigned<T>) {
    if (assign) {
      Object.assign(comp, assign);
    }

    return comp.load();
  }

  abstract prepare(bufferCursor: BufferCursor): void;

  appendRaw(raw: Buffer | number) {
    this.raw = Buffer.concat([
      this.raw,
      typeof raw === "number" ? Buffer.from([raw]) : raw,
    ]);
    this._raw = this.raw.toString();

    return this;
  }

  rejectUnexpectedTokenError(options?: RejectOptions): never {
    const err = new UnexpectedTokenError(
      this.filename,
      this.body,
      options?.position ?? this.bufferCursor.position,
      {
        message: options?.message ?? "Unexpected token",
      }
    );

    Error.captureStackTrace(err, Base.prototype.rejectUnexpectedTokenError);

    throw err;
  }

  lastChildren() {
    return this.children[this.children.length - 1];
  }

  removeChildren(children: Base) {
    this.children = this.children.filter((elm) => elm !== children);
  }

  toJSON() {
    return <BaseExportTypeJSON>{
      $type: this.$type,
      pos: this.pos,
      end: this.end,
      children: this.children.length ? (this.children as any[]) : undefined,
    };
  }

  static serialize(opts: BaseType): Buffer {
    return Buffer.from([]);
  }
}
