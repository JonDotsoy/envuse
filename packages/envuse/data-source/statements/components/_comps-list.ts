export { Block } from "./block";
export { CommentInline as Comment } from "./comment-inline";
export { CommentOperator } from "./comment-operator";
export { CommentOperatorStatement } from "./comment-operator-statement";
export { Space } from "./space";
export { SpaceNewLine } from "./space-new-line";
export { StatementObject } from "./statement-object";
export { SymbolEqual } from "./symbol-equal";
export { Variable } from "./variable";
export { VariableKey } from "./variable-key";
export { VariableValue } from "./variable-value";
