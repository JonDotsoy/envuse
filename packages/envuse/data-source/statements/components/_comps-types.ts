import { BaseSerializeOption } from "../tdo/base-serialize-option";
import { Block } from "./block";

export type BlockObj = BaseSerializeOption<Block>;
