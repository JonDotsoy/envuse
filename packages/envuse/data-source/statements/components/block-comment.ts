import { ArrCursor, Validator } from "../lib/arr-cursor";
import { BufferCursor } from "../lib/buffer-cursor";
import { CharactersKey } from "../tdo/characters-key";
import { TypeNamesList } from "../tdo/type-names-list";
import { Base } from "./base";

export type BlockCommentType = {
  $type: "BlockComment";
  value: string;
  [k: string]: any;
};

export class BlockComment extends Base {
  $type = "BlockComment" as const;
  buff: number[] = [];

  get value(): string {
    return this.body.slice(this.pos, this.end).toString();
  }

  prepare(bufferCursor: BufferCursor): void {
    bufferCursor.forward(3);

    const validatorSpaceOrNewLine: Validator<BufferCursor> = (
      cursor,
      actions
    ) => {
      cursor.forward();

      if (cursor.current() === CharactersKey.space) {
        while (cursor.current() === CharactersKey.space) {
          cursor.forward();
        }
      }

      if (cursor.current() === CharactersKey.newLineLF) return actions.continue;
    };

    while (bufferCursor.has()) {
      let matched;
      if (
        bufferCursor.isClosed() ||
        (Buffer.from(Array(3).fill(CharactersKey.numberSign)).equals(
          Buffer.from(bufferCursor.currentAndPrev(3))
        ) &&
          (([matched] = bufferCursor.clone().match(validatorSpaceOrNewLine)),
          matched))
      ) {
        bufferCursor.forward();
        return;
      }

      this.buff.push(bufferCursor.current());
      bufferCursor.forward();
    }
  }

  toJSON() {
    return {
      ...super.toJSON(),
      // value: this.value,
    };
  }

  static serialize(blockComment: BlockCommentType) {
    return Buffer.concat([Buffer.from(blockComment.value), Buffer.from("\n")]);
  }
}
