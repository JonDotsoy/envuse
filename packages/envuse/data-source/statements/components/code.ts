import { Base } from "./base";

export class Code extends Base {
  $type = "Code" as const;

  prepare() {
    // let initialized = false
    // while (true) {
    //   const { done, value } = gen.next();
    //   if (done || !value)
    //     return;
    //   const [index, current_char, { next }] = value;
    //   if (!initialized && current_char.equals(Buffer.from([0x20]))) {
    //     if (next(2).equals(Buffer.from([0x20, 0x20]))) continue;
    //     initialized = true;
    //     this.pos = index;
    //     continue;
    //   }
    //   if (
    //     current_char.equals(Buffer.from([0x0a]))
    //   ) {
    //     if (next(2).equals(Buffer.from([0x0a, 0x23]))) {
    //       gen.next()
    //       continue;
    //     }
    //     return;
    //   }
    //   this.appendRaw(current_char);
    // }
  }
}
