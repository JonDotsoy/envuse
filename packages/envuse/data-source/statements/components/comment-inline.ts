import { Base } from "./base";
import { BufferCursor } from "../lib/buffer-cursor";
import { BCharType } from "../tdo/b-char-type";
import { BaseSerializeOption } from "../tdo/base-serialize-option";
import { Space } from "./space";
import { charactersKeys } from "../tdo/characters-keys";
import { CharactersKey } from "../tdo/characters-key";

export type CommentType = {
  $type: "Comment";
  value: string;
  [k: string]: any;
};

export class CommentInline extends Base {
  $type = "Comment" as const;

  #valueRaw: number[] = [];

  get value() {
    return Buffer.from(this.#valueRaw).toString();
  }

  prepare(bufferCursor: BufferCursor<BCharType>): void {
    bufferCursor.forward();

    if (bufferCursor.current() === CharactersKey.space) {
      this.createElement(Space);
    }

    while (bufferCursor.has()) {
      if (
        bufferCursor.currentIs(CharactersKey.newLineLF) ||
        bufferCursor.isClosed()
      ) {
        bufferCursor.forward();
        return;
      }

      this.#valueRaw.push(bufferCursor.current());
      bufferCursor.forward();
    }
  }

  toJSON() {
    return {
      ...super.toJSON(),
      value: this.value,
    };
  }

  static serialize(comp: CommentType) {
    return Buffer.from(`# ${comp.value}\n`);
  }
}
