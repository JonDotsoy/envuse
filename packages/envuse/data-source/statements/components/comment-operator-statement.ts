import { Base } from "./base";
import { BufferCursor } from "../lib/buffer-cursor";
import { BCharType } from "../tdo/b-char-type";
import { Space } from "./space";
import { CharactersKey as K } from "../tdo/characters-key";
import { StatementObject, StatementObjectType } from "./statement-object";
import { BaseSerializeOption } from "../tdo/base-serialize-option";

export type CommentOperatorStatementType = {
  $type: "CommentOperatorStatement";
  statements: StatementObjectType[];
};

export class CommentOperatorStatement extends Base {
  $type = "CommentOperatorStatement" as const;
  statements: StatementObject[] = [];

  prepare(bufferCursor: BufferCursor<BCharType>): void {
    while (true) {
      if (
        bufferCursor.current() === K.newLineLF ||
        bufferCursor.current() === K.numberSign ||
        bufferCursor.isClosed()
      ) {
        bufferCursor.forward();
        return;
      }

      this.statements.push(this.createElement(StatementObject));

      if (bufferCursor.current() === K.newLineLF) {
        bufferCursor.forward();
        return;
      }

      if (bufferCursor.current() === K.space) {
        this.createElement(Space);
      }
    }
  }

  toJSON() {
    return {
      ...super.toJSON(),
      statements: this.statements,
    };
  }

  static serialize(comp: CommentOperatorStatementType) {
    return Buffer.concat(
      comp.statements.map((s, i, { length }) =>
        Buffer.concat([
          StatementObject.serialize(s),
          Buffer.from(length - 1 === i ? "" : " "),
        ])
      )
    );
  }
}
