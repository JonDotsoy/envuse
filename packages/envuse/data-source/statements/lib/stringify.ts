import { BlockType } from "../components/block";
import * as comps from "../components/_comps-list";
import { stringifyCtx } from "./stringify-ctx";
import { StringifyOptions } from "./stringify-options";

const serialize_comp = (comp: BlockType) => {
  switch (comp.$type) {
    case "Block":
      return comps.Block.serialize(comp);
    default:
      throw new Error(
        `No possible serialize component with type ${comp.$type}`
      );
  }
};

export const stringify = (comp: BlockType, options?: StringifyOptions) => {
  stringifyCtx.start(options);
  const res = serialize_comp(comp);
  stringifyCtx.end();
  return res;
};
