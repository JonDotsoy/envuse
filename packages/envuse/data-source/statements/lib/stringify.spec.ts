import { stringify } from "./stringify";
import { b } from "./to-buffer";
import { DataSource } from "../../data-source";
import { BlockType } from "../components/block";

describe("parse ast", () => {
  it("should parse ast variable", () => {
    const body = DataSource.createDataSource(Buffer.concat([b(`a=3\n`)]));

    expect(JSON.stringify(body, null, 2)).toMatchSnapshot();
  });

  it("should parse ast operator", () => {
    const body = DataSource.createDataSource(
      Buffer.concat([b(`#; if true\n`), b(`a=b\n`), b(`#; fi\n`)])
    );

    expect(JSON.stringify(body, null, 2)).toMatchSnapshot();
  });
});

describe("serialize", () => {
  it("serialize component block", () => {
    const block: BlockType = {
      $type: "Block",
      children: [
        {
          $type: "Variable",
          keyVariable: {
            $type: "VariableKey",
            value: "aa",
          },
          valueVariable: {
            $type: "VariableValue",
            value: "bbb",
          },
        },
      ],
    };

    expect(stringify(block).toString()).toMatchInlineSnapshot(`
    "aa : string = \\"bbb\\"
    "
    `);
  });

  it("serialize component variable and comment", () => {
    const block: BlockType = {
      $type: "Block",
      children: [
        {
          $type: "Comment",
          value: "Iam comment",
        },
        {
          $type: "Variable",
          keyVariable: {
            $type: "VariableKey",
            value: "aaa",
          },
          valueVariable: {
            $type: "VariableValue",
            value: "bbb",
          },
        },
      ],
    };

    expect(stringify(block).toString()).toMatchInlineSnapshot(`
    "# Iam comment
    aaa : string = \\"bbb\\"
    "
    `);
  });

  it("serialize component variable and comment", () => {
    const block = DataSource.createDataSource(
      Buffer.concat([b(`#; if true === true\n`), b(`aaa=bbb\n`), b(`#; fi\n`)])
    );

    expect(stringify(block).toString()).toMatchInlineSnapshot(`
    "#; if true === true
    aaa : string = \\"bbb\\"
    #; fi
    "
    `);
  });

  it("should serialize component complex", () => {
    const block = DataSource.createDataSource(
      Buffer.concat([
        b(`# single comment\n`),
        b(`abc=abc\n`),
        b(`c="abnc\\nasd"\n`),
        b(`#;if true === 12 === abc.def === "abc"\n`),
        b(`cde=321\n`),
        b(`#;fi\n`),
        b(`\n`),
        b(`# Comment\n`),
        b(`other=true\n`),
      ])
    );

    expect(stringify(block).toString()).toMatchInlineSnapshot(`
      "# single comment
      abc : string = \\"abc\\"
      c : string = \\"abnc\\\\nasd\\"
      #; if true === 12 === abc.def === \\"abc\\"
      cde : string = \\"321\\"
      #; fi
      # Comment
      other : string = \\"true\\"
      "
    `);
  });

  it("should serialize component declaration manual", () => {
    const block: BlockType = {
      $type: "Block",
      children: [
        {
          $type: "Comment",
          value: "I am comment",
        },
        {
          $type: "CommentOperator",
          operator: {
            $type: "VariableKey",
            value: "if",
          },
          statement: {
            $type: "CommentOperatorStatement",
            statements: [
              {
                $type: "StatementObject",
                type: "Boolean",
                value: true,
              },
            ],
          },
          block: {
            $type: "Block",
            children: [],
          },
        },
      ],
    };

    expect(stringify(block).toString()).toMatchInlineSnapshot(`
      "# I am comment
      #; if true
      #; fi
      "
    `);
  });
});
