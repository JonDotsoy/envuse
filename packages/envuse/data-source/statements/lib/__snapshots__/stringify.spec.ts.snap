// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse ast should parse ast operator 1`] = `
"{
  \\"$type\\": \\"Block\\",
  \\"pos\\": 0,
  \\"end\\": 21,
  \\"children\\": [
    {
      \\"$type\\": \\"CommentOperator\\",
      \\"pos\\": 0,
      \\"end\\": 20,
      \\"operator\\": {
        \\"$type\\": \\"VariableKey\\",
        \\"pos\\": 3,
        \\"end\\": 5,
        \\"value\\": \\"if\\"
      },
      \\"statement\\": {
        \\"$type\\": \\"CommentOperatorStatement\\",
        \\"pos\\": 6,
        \\"end\\": 11,
        \\"statements\\": [
          {
            \\"$type\\": \\"StatementObject\\",
            \\"pos\\": 6,
            \\"end\\": 10,
            \\"type\\": \\"Boolean\\",
            \\"value\\": true
          }
        ]
      },
      \\"block\\": {
        \\"$type\\": \\"Block\\",
        \\"pos\\": 11,
        \\"end\\": 20,
        \\"children\\": [
          {
            \\"$type\\": \\"Variable\\",
            \\"pos\\": 11,
            \\"end\\": 15,
            \\"keyVariable\\": {
              \\"$type\\": \\"VariableKey\\",
              \\"pos\\": 11,
              \\"end\\": 12,
              \\"value\\": \\"a\\"
            },
            \\"valueVariable\\": {
              \\"$type\\": \\"VariableValue\\",
              \\"pos\\": 13,
              \\"end\\": 14,
              \\"value\\": \\"b\\"
            }
          }
        ]
      }
    },
    {
      \\"$type\\": \\"SpaceNewLine\\",
      \\"pos\\": 20,
      \\"end\\": 21
    }
  ]
}"
`;

exports[`parse ast should parse ast variable 1`] = `
"{
  \\"$type\\": \\"Block\\",
  \\"pos\\": 0,
  \\"end\\": 4,
  \\"children\\": [
    {
      \\"$type\\": \\"Variable\\",
      \\"pos\\": 0,
      \\"end\\": 4,
      \\"keyVariable\\": {
        \\"$type\\": \\"VariableKey\\",
        \\"pos\\": 0,
        \\"end\\": 1,
        \\"value\\": \\"a\\"
      },
      \\"valueVariable\\": {
        \\"$type\\": \\"VariableValue\\",
        \\"pos\\": 2,
        \\"end\\": 3,
        \\"value\\": \\"3\\"
      }
    }
  ]
}"
`;
